<a href="https://colab.research.google.com/github/Leima20/Sprint_8_Leidy_Pimentel/blob/main/Sprint_8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

NIVEL 1

Exercici 1: Exploració inicial del dataset
🎯 Objectiu: entendre l’estructura general del dataset i detectar possibles problemes abans de començar a netejar o transformar dades.
En aquest exercici has de realitzar una exploració inicial del DataFrame per respondre, com a mínim, a les següents preguntes:

» Quantes files i columnes té el dataset?
» Quin tipus de dades conté cada columna?
» Hi ha files duplicades?
» Les dades semblen coherents a primera vista?
» Detectes possibles problemes o inconsistències evidents?
Per fer-ho, hauràs de:

» Analitzar les dimensions del dataset
» Mostrar una mostra representativa de files
» Revisar els tipus de variables
» Les dades semblen coherents a primera vista?
» Detectar files duplicades
» Identificar i descriure possibles problemes inicials (sense solucionar-los encara)

Introducción

En este ejercicio realizamos la carga y exploración básica del dataset de empleados para verificar su integridad, estructura y calidad antes de proceder a la limpieza.

Importaciones

In [5]:
import pandas as pd

Importamos el dataframe

In [6]:
file_path = 'sprint8_complex.xlsx'
df = pd.read_excel(file_path)

df

,Dades de l'enquesta de treballadors 2025,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5,Unnamed: 6,Unnamed: 7,Unnamed: 8,Unnamed: 9,Unnamed: 10,Unnamed: 11,Unnamed: 12,Unnamed: 13,Unnamed: 14,Unnamed: 15,Unnamed: 16,Unnamed: 17,Unnamed: 18,Unnamed: 19
0,Generades aleatòriament per seguretat,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,CÃ rrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
3,0,Joana,Gil Navarro,66722344X,Espanya,Valladolid,23.0,3.0,1958.0,Dona,"1,469 €",True,NaN,Grup B,Cap de projecte,3.0,True,32108.0,25.0,10.1
4,1,Marc,Muñoz,48840994W,Espanya,Alacant,8.0,11.0,1960.0,H,"2,718 €",NaN,True,Grup C,Senior analyst,0.0,True,19496.0,10.4,18.7
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1005,1002,Alba,Muñoz,97954019T,Espanya,Bilbao,25.0,10.0,1983.0,D,"1,409 €",NaN,True,Grup B,Responsable de vendes,0.0,NaN,NaN,NaN,14.7
1006,1003,Claudia,Gil Ferrer,55297384F,Espanya,València,1.0,11.0,1981.0,H,1.789 €,NaN,True,Grup B,Responsable de vendes,0.0,True,29952.0,9.2,20.3
1007,1004,Oriol,Ferrer,52590001R,Espanya,Zaragoza,15.0,7.0,1973.0,D,786€,NaN,True,Grup A,Analista,0.0,True,81381.0,6.5,13.9
1008,1005,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


**Filas y columnas del dataset**

Se emplea 'df.shape' para conocer las dimensiones del dataset. l dataset está compuesto por **1010 filas y 20 columnas.** Esta información proporciona una visión general del volumen de datos disponible y sirve como punto de partida para el análisis exploratorio.

In [7]:

filas, columnas = df.shape
print(f" El dataset tiene {filas} filas y {columnas} columnas ")

 El dataset tiene 1010 filas y 20 columnas 


**Tipos de datos que contiene cada columna**

Todas las columnas del dataset se han cargado con el tipo de dato object. Esto indica que Pandas ha interpretado todas las variables como texto, por lo que será necesario revisar si existen columnas que deberían convertirse a otros tipos de datos, como valores numéricos o fechas, para facilitar su análisis.

In [8]:
print("\n--- Tipos de datos ---")
print(df.info())


--- Tipos de datos ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1010 entries, 0 to 1009
Data columns (total 20 columns):
 #   Column                                    Non-Null Count  Dtype 
---  ------                                    --------------  ----- 
 0   Dades de l'enquesta de treballadors 2025  1008 non-null   object
 1   Unnamed: 1                                1007 non-null   object
 2   Unnamed: 2                                1007 non-null   object
 3   Unnamed: 3                                1007 non-null   object
 4   Unnamed: 4                                1007 non-null   object
 5   Unnamed: 5                                1007 non-null   object
 6   Unnamed: 6                                1007 non-null   object
 7   Unnamed: 7                                1007 non-null   object
 8   Unnamed: 8                                1007 non-null   object
 9   Unnamed: 9                                967 non-null    object
 10  Unnamed: 10             

**Duplicados**

Se utiliza el método duplicated().sum() para identificar el número de filas completamente duplicadas en el dataset.

Los resultados muestran que el dataset no contiene filas completamente duplicadas, ya que se han detectado 0 registros duplicados.

In [9]:
duplicados = df.duplicated().sum()
print(f"el Dataset contiene {duplicados} filas duplicadas")

el Dataset contiene 0 filas duplicadas


**Coherencia de los datos**

La muestra aleatoria permite comprobar que el dataset corresponde a un conjunto de datos de empleados y que, en general, mantiene una estructura consistente. Sin embargo, durante la inspección visual se observan algunos indicios de posibles inconsistencias que deberán analizarse con mayor detalle en los siguientes apartados.

Al revisar las primeras filas del dataset `display(df.head())` a primera vista, el contenido parece corresponder a información de empleados. Sin embargo, se observa que las tres primeras filas no contienen datos del dataset, sino información adicional del archivo (como un título y filas vacías). Asimismo, los nombres de las variables no se han interpretado correctamente, apareciendo columnas con nombres como **Unnamed**. Esto indica que el archivo no se ha leído con la fila de encabezados adecuada, aspecto que será necesario corregir mas adelante.

In [10]:
display(df.head())

,Dades de l'enquesta de treballadors 2025,Unnamed: 1,Unnamed: 2,Unnamed: 3,Unnamed: 4,Unnamed: 5,Unnamed: 6,Unnamed: 7,Unnamed: 8,Unnamed: 9,Unnamed: 10,Unnamed: 11,Unnamed: 12,Unnamed: 13,Unnamed: 14,Unnamed: 15,Unnamed: 16,Unnamed: 17,Unnamed: 18,Unnamed: 19
0,Generades aleatòriament per seguretat,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,CÃ rrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
3,0,Joana,Gil Navarro,66722344X,Espanya,Valladolid,23.0,3.0,1958.0,Dona,"1,469 €",True,NaN,Grup B,Cap de projecte,3.0,True,32108.0,25.0,10.1
4,1,Marc,Muñoz,48840994W,Espanya,Alacant,8.0,11.0,1960.0,H,"2,718 €",NaN,True,Grup C,Senior analyst,0.0,True,19496.0,10.4,18.7


**Posibles problemas o inconsistencias evidentes**

Se han encontrado algunas inconsistencias al revisar el dataset:

 * Los nombres de los encabezados de las columnas no se han interpretado correctamente debido a la estructura del archivo.
 * Se observan inconsistencias en el formato de la columna de salarios, por lo que será necesario revisar su conversión a un tipo numérico.
 * La variable Género presenta una falta de estandarización, ya que contiene valores como Dona, Home, D, H, F, M, NC, A, ?? y valores nulos.
 * Se detectan problemas de codificación (encoding), reflejados en caracteres mal representados, como CÃrrec.
 * Existe una posible redundancia estructural entre las variables Fills, No Fills y Nombre_fills, por lo que será necesario comprobar si contienen información duplicada o inconsistente.

Exercici 2:

 Validacions bàsiques de les dades i decisions d'eliminació
🎯 Objectiu: assegurar que les dades compleixen unes condicions mínimes de qualitat i determinar quines files poden continuar sent analitzades i quines s'han de descartar.
En aquest exercici hauràs de validar aspectes clau del dataset que són crítics en qualsevol context real d'anàlisi de dades, prenent decisions justificades sobre la validesa o eliminació de registres.

Aspectes a validar:

» Existeixen identificadors únics (DNI) per a cada persona?

» Què faràs amb les files duplicades o registres repetits?

» Hi ha valors obligatoris que falten o informació incoherent?

» Detectes formats incorrectes, errors de codificació o valors impossibles?

Concretament, hauràs de:

» Corregir possibles problemes de codificació i accentuació en els noms de les columnes i en els valors de text, si escau.

» Detectar i eliminar files duplicades exactes, indicant quants registres s'han eliminat.

» Comprovar la unicitat del DNI. En cas de detectar DNIs duplicats, hauràs d'analitzar la coherència de la informació associada i descartar o anul·lar els registres que no siguin fiables, justificant clarament la decisió presa.

» Identificar els camps clau que no haurien de contenir valors nuls, detectar registres amb informació insuficient i eliminar únicament aquelles files completament buides o amb informació irrecuperable o no substituïble, explicant per què no és possible ni raonable imputar-les.

Importamos las librerias necesarias

In [11]:
import pandas as pd
import re
from collections import defaultdict



Leemos el fichero excel y lo cargamos en un DataFrame de pandas. Ignoramos las 2 primeras filas y usamos la tercera fila como cabecera.
Además, establecemos la primera columna como índice del DataFrame.


In [12]:
df = pd.read_excel("sprint8_complex.xlsx", header=3, index_col=0)


Definimos una función auxiliar para reparar los errores de codificación.

Al inspeccionar el dataset se detectan errores de codificación tanto en los nombres de algunas columnas como en varios valores de texto. Algunos ejemplos son `Cà rrec` en lugar de `Càrrec`, `Adrià ` en lugar de `Adrià` o `Muñoz` en lugar de `Muñoz`.

Para corregir estos problemas, se define una función que intenta reparar los textos mediante una conversión entre las codificaciones CP1252 y UTF-8. Si la conversión no es posible, se devuelve el valor original para evitar modificar información que ya sea correcta.

In [13]:
def reparar_codificacion(texto) -> str:
    """
    Corrige posibles errores de codificación en una cadena de texto.

    Intenta reconstruir caracteres mal codificados convirtiendo desde
    Windows-1252 (cp1252) a UTF-8. Si la conversión no es posible,
    devuelve el texto original.

    Args:
        texto (str): Texto a reparar.

    Returns:
        str: Texto reparado o el original si no requiere corrección.
    """

    try:
        return texto.encode("cp1252").decode("utf-8")

    except (UnicodeEncodeError, UnicodeDecodeError):
        return texto

 **Corrección de la codificación en las cabeceras**

Una vez definida la función de reparación, se aplica a los nombres de las columnas del dataset.

Posteriormente, se comparan las cabeceras originales con las reparadas para identificar cuáles presentaban problemas de codificación. Si se detectan diferencias, se actualizan los nombres de las columnas utilizando la versión corregida.

De esta forma se garantiza que los nombres de las variables sean legibles y puedan utilizarse correctamente durante el resto del análisis.

In [14]:
# Guardar las cabeceras originales
cabeceras_originales = df.columns

print(f"Cabeceras originales: {cabeceras_originales.tolist()}")

Cabeceras originales: ['Nom', 'Cognoms', 'DNI', "País d'origen", 'Ciutat', 'Dia de Naixement', 'Mes de Naixement', 'Any de Naixement', 'Gènere', 'Salari mensual', 'Fills', 'No Fills', 'Grup Professional', 'CÃ\xa0rrec', 'Nombre_fills', 'Te_cotxe', 'Km_anuals', 'Consum_mitja_L_100km', 'Temperatura_mitjana_ciutat']


In [15]:
# Reparar la codificación de las cabeceras
cabeceras_reparadas = cabeceras_originales.map(
    lambda columna: (
        reparar_codificacion(columna)
        if isinstance(columna, str)
        else columna
    )
)

In [16]:
# Identificar las cabeceras que han sido modificadas
cabeceras_modificadas = cabeceras_originales[
    cabeceras_originales != cabeceras_reparadas
]

In [17]:
# Sustituir las cabeceras si se han encontrado errores
if len(cabeceras_modificadas) > 0:

    print(
        "Se han reparado las siguientes cabeceras debido "
        "a problemas de codificación:"
    )

    print(cabeceras_modificadas.tolist())

    df.columns = cabeceras_reparadas

else:

    print(
        "No se han detectado problemas de codificación "
        "en las cabeceras."
    )

Se han reparado las siguientes cabeceras debido a problemas de codificación:
['CÃ\xa0rrec']


Cabecera reparada

In [18]:
df.columns.tolist()

['Nom',
 'Cognoms',
 'DNI',
 "País d'origen",
 'Ciutat',
 'Dia de Naixement',
 'Mes de Naixement',
 'Any de Naixement',
 'Gènere',
 'Salari mensual',
 'Fills',
 'No Fills',
 'Grup Professional',
 'Càrrec',
 'Nombre_fills',
 'Te_cotxe',
 'Km_anuals',
 'Consum_mitja_L_100km',
 'Temperatura_mitjana_ciutat']

**Corrección de la codificación en los valores de texto**

Una vez corregidas las cabeceras, se aplica el mismo procedimiento a todas las columnas de tipo texto del dataset.

Para cada columna se comparan los valores originales con los valores reparados. Si se detectan diferencias, únicamente se sustituyen aquellos registros afectados por problemas de codificación, manteniendo el resto de la información sin modificaciones.

Además, se informa del número de valores corregidos en cada columna para documentar las transformaciones realizadas durante el proceso de limpieza de datos.

In [19]:
# Revisar todas las columnas de tipo texto
for columna in df.select_dtypes(include=["object", "string"]):

    # Guardar la serie original
    serie_original = df[columna]

    # Aplicar la reparación únicamente a los valores de texto
    serie_reparada = serie_original.map(
        lambda valor: (
            reparar_codificacion(valor)
            if isinstance(valor, str)
            else valor
        )
    )

    # Contar cuántos valores han cambiado
    valores_reparados = (serie_original != serie_reparada).sum()

    # Sustituir la columna únicamente si se han realizado cambios
    if valores_reparados > 0:

        print(
            f"Se han reparado {valores_reparados} valores "
            f"en la columna '{columna}'."
        )

        df[columna] = serie_reparada

Se han reparado 129 valores en la columna 'Nom'.
Se han reparado 341 valores en la columna 'Cognoms'.
Se han reparado 1 valores en la columna 'DNI'.
Se han reparado 195 valores en la columna 'País d'origen'.
Se han reparado 161 valores en la columna 'Ciutat'.
Se han reparado 41 valores en la columna 'Gènere'.
Se han reparado 843 valores en la columna 'Salari mensual'.
Se han reparado 588 valores en la columna 'Fills'.
Se han reparado 422 valores en la columna 'No Fills'.
Se han reparado 2 valores en la columna 'Grup Professional'.
Se han reparado 131 valores en la columna 'Càrrec'.
Se han reparado 349 valores en la columna 'Te_cotxe'.


**Detección y eliminación de filas duplicadas exactas**

Antes de continuar con las validaciones, se comprueba la existencia de registros duplicados exactos.

Se consideran filas duplicadas aquellas que presentan exactamente los mismos valores en todas sus columnas. Estos registros no aportan información adicional y podrían alterar los resultados del análisis al contabilizar varias veces un mismo empleado.

Por este motivo, se conservan únicamente las primeras apariciones y se eliminan las copias duplicadas.

In [20]:
# Identificar todos los registros que forman parte de una duplicidad
filas_duplicadas = df[df.duplicated(keep=False)].sort_values("DNI")

if not filas_duplicadas.empty:

    print(
        "Se han detectado los siguientes registros duplicados exactos:\n"
    )

    display(filas_duplicadas)

else:

    print("No se han encontrado filas duplicadas exactas.")

Se han detectado los siguientes registros duplicados exactos:



,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,Càrrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
757,Oriol,Ferrer,52590001R,Espanya,Zaragoza,15.0,7.0,1973.0,D,786€,NaN,True,Grup A,Analista,0.0,True,81381.0,6.5,13.9
1004,Oriol,Ferrer,52590001R,Espanya,Zaragoza,15.0,7.0,1973.0,D,786€,NaN,True,Grup A,Analista,0.0,True,81381.0,6.5,13.9
464,Ida,Karlsen Eriksen,55283701C,Noruega,Oslo,2.0,9.0,1992.0,D,"1,614 €",True,NaN,Grup B,Analista,3.0,NaN,NaN,NaN,10.6
1001,Ida,Karlsen Eriksen,55283701C,Noruega,Oslo,2.0,9.0,1992.0,D,"1,614 €",True,NaN,Grup B,Analista,3.0,NaN,NaN,NaN,10.6
730,Claudia,Gil Ferrer,55297384F,Espanya,València,1.0,11.0,1981.0,H,1.789 €,NaN,True,Grup B,Responsable de vendes,0.0,True,29952.0,9.2,20.3
1003,Claudia,Gil Ferrer,55297384F,Espanya,València,1.0,11.0,1981.0,H,1.789 €,NaN,True,Grup B,Responsable de vendes,0.0,True,29952.0,9.2,20.3
372,Alba,Muñoz,97954019T,Espanya,Bilbao,25.0,10.0,1983.0,D,"1,409 €",NaN,True,Grup B,Responsable de vendes,0.0,NaN,NaN,NaN,14.7
1002,Alba,Muñoz,97954019T,Espanya,Bilbao,25.0,10.0,1983.0,D,"1,409 €",NaN,True,Grup B,Responsable de vendes,0.0,NaN,NaN,NaN,14.7


In [21]:
# Eliminar los registros duplicados conservando la primera aparición
filas_iniciales = len(df)

df.drop_duplicates(inplace=True)

filas_eliminadas = filas_iniciales - len(df)

print(f"\nSe han eliminado {filas_eliminadas} filas duplicadas exactas.")


Se han eliminado 4 filas duplicadas exactas.


## Validación de la unicidad del DNI

El DNI se considera el identificador único de cada empleado, por lo que no deberían existir dos registros con el mismo valor.

Para comprobar su unicidad, se identifican todos los DNIs repetidos y se revisa la información asociada a cada uno de ellos.

En este dataset, los registros que comparten un mismo DNI presentan información inconsistente y no es posible determinar cuál de ellos es el correcto. Ante la imposibilidad de decidir de forma objetiva qué registro conservar, se opta por eliminar todos los registros asociados a los DNIs duplicados.

Esta decisión evita mantener información potencialmente errónea que pueda afectar a los análisis posteriores.

In [22]:
# Identificar todos los registros con un DNI duplicado
duplicados = df[df.duplicated(subset=["DNI"], keep=False)]

if duplicados.empty:

    print("No se han encontrado DNIs duplicados.")

else:

    for dni, grupo in duplicados.groupby("DNI"):

        print(f"\nRegistros con el DNI duplicado {dni}:")

        display(grupo)

        # Eliminar todos los registros asociados a ese DNI
        df.drop(index=grupo.index, inplace=True)

        print(
            f"Se han eliminado {len(grupo)} registros "
            f"con el DNI {dni}."
        )


Registros con el DNI duplicado 15909601K:


,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,Càrrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
250,Laia,Navarro Vila,15909601K,Espanya,Bilbao,30.0,5.0,1958.0,D,1235 euros,True,True,Grup D,Analista junior,0.0,False,NaN,NaN,10.3
604,Laia,Navarro,15909601K,Espanya,Bilbao,30.0,5.0,1958.0,D,"1,676 €",True,False,Grup A,Analista,1.0,True,50.0,NaN,14.2


Se han eliminado 2 registros con el DNI 15909601K.

Registros con el DNI duplicado 16618423V:


,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,Càrrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
654,Oriol,Gómez,16618423V,Espanya,Madrid,10.0,1.0,1968.0,Home,1731 euros,True,NaN,Grup B,Data analyst,4.0,NaN,NaN,NaN,NaN
754,Oriol,Gómez,16618423V,Espanya,Barcelona,10.0,1.0,1968.0,Home,2.358 €,True,NaN,NaN,NaN,4.0,NaN,NaN,NaN,NaN
758,Oriol,Gómez,16618423V,Espanya,Màlaga,10.0,1.0,1968.0,H,1879 euros,True,NaN,Grup B,Data Analyst,4.0,False,NaN,NaN,NaN
759,Oriol,Gómez,16618423V,Espanya,Sevilla,10.0,1.0,1968.0,NaN,"2,000 €",NaN,NaN,Grup A,Responsable de vendes,5.0,NaN,NaN,NaN,20.7


Se han eliminado 4 registros con el DNI 16618423V.

Registros con el DNI duplicado 49529438G:


,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,Càrrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
558,Alexia,Pérez Díaz,49529438G,Espanya,Màlaga,12.0,10.0,1981.0,Dona,"1,176 €",NaN,True,Grup A,Data Analyst,0.0,False,NaN,NaN,38.0
692,Alexia,Pérez Díaz,49529438G,Espanya,Màlaga,12.0,10.0,1981.0,Dona,1663€,NaN,True,Grup C,Cap de projecte,NaN,True,80000.0,10.7,16.9


Se han eliminado 2 registros con el DNI 49529438G.

Registros con el DNI duplicado 54013571W:


,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,Càrrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
89,Guillem,Pujol Costa,54013571W,Espanya,A Coruña,17.0,7.0,1983.0,D,1.574 €,True,NaN,Grup B,Project lead,4.0,False,NaN,NaN,14.6
142,Marta,Ruiz,54013571W,Espanya,Palma,24.0,5.0,1982.0,H,"1,276 €",True,True,Grup A,Analista,0.0,False,NaN,NaN,15.7


Se han eliminado 2 registros con el DNI 54013571W.

Registros con el DNI duplicado 90585882A:


,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,Càrrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
25,Sara,Serra,90585882A,Espanya,Tarragona,28.0,5.0,2002.0,D,993 €,NaN,True,Grup A,Administratiu,0.0,False,NaN,NaN,15.8
104,Oriol,Vila Casals,90585882A,Espanya,Madrid,30.0,9.0,1964.0,H,1987 euros,NaN,True,Grup B,Responsable de vendes,0.0,False,NaN,NaN,13.5
114,Pol,Serra,90585882A,Espanya,Girona,4.0,12.0,1960.0,H,"1,178 €",NaN,True,Grup A,Data Analyst,NaN,False,NaN,NaN,16.2
281,Xavier,Vidal,90585882A,Espanya,Alacant,2.0,7.0,1958.0,H,1179 euros,NaN,NaN,Grup A,Tècnic IT,0.0,True,63574.0,6.0,16.8
913,Sara,Serra,90585882A,Espanya,Tarragona,28.0,5.0,2002.0,Dona,"1,000 €",NaN,True,Grup A,Administratiu,0.0,False,NaN,NaN,15.8
1000,Sara,Serra,90585882A,Espanya,Tarragona,28.0,5.0,2002.0,D,"1,073 €",NaN,True,Grup A,Administratiu,0.0,False,NaN,NaN,15.8


Se han eliminado 6 registros con el DNI 90585882A.


verificacion despues de eliminar duplicados

Comprobacion de valores nulos

In [23]:
# Comprobar si siguen existiendo DNIs duplicados después de la limpieza

duplicados_despues = df[df.duplicated(subset=["DNI"], keep=False)]

if duplicados_despues.empty:

    print("No quedan DNIs duplicados en el dataset.")

else:

    print("Todavía existen DNIs duplicados:")
    display(duplicados_despues)

No quedan DNIs duplicados en el dataset.


Revisar valores nulos

In [24]:
# Revisar valores nulos por columna

valores_nulos = df.isnull().sum()

# Mostrar solamente columnas con valores faltantes

nulos = valores_nulos[valores_nulos > 0]

display(nulos)


,0
Nom,1
Cognoms,1
DNI,1
País d'origen,1
Ciutat,1
Dia de Naixement,1
Mes de Naixement,1
Any de Naixement,1
Gènere,40
Salari mensual,1


Se detectan valores nulos en varias columnas del dataset. Además, durante la inspección de los registros afectados se observa un registro completamente vacío (índice 1005) y otro con valores aparentemente inválidos (índice 1006), como ???, XXXX o Narnia. La validación de estos registros se realizará en los siguientes bloques, donde se comprobarán los tipos de datos, los rangos y los valores permitidos.

In [25]:
# Mostrar registros con algún valor nulo

filas_nulas = df[df.isnull().any(axis=1)]

display(filas_nulas)

,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,Càrrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
0,Joana,Gil Navarro,66722344X,Espanya,Valladolid,23.0,3.0,1958.0,Dona,"1,469 €",True,NaN,Grup B,Cap de projecte,3.0,True,32108.0,25.0,10.1
1,Marc,Muñoz,48840994W,Espanya,Alacant,8.0,11.0,1960.0,H,"2,718 €",NaN,True,Grup C,Senior analyst,0.0,True,19496.0,10.4,18.7
2,Noa,Serra,14308421X,Espanya,Alacant,27.0,4.0,1961.0,D,1358 euros,True,NaN,Grup A,Tècnic IT,4.0,NaN,NaN,NaN,16.7
3,Pol,Gil,58586340F,Espanya,Sevilla,13.0,10.0,1985.0,H,"1,478 €",True,NaN,Grup B,Data Analyst,2.0,False,NaN,NaN,18.3
4,David,Vila,82070937P,Espanya,Bilbao,29.0,11.0,1965.0,Dona,"1,284 €",NaN,True,Grup B,Administratiu,NaN,True,NaN,11.8,13.1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
998,Chloé,Dubois,66354268T,França,Nice,2.0,12.0,1958.0,H,"1,954 €",True,NaN,Grup B,Data Analyst,1.0,NaN,NaN,NaN,11.7
999,Adrià,Vila Díaz,57511543T,Espanya,Palma,31.0,12.0,1991.0,H,"2,010 €",True,NaN,Grup C,Analista,1.0,False,NaN,NaN,-10.0
651,Kasper,Pettersen,62042563B,Noruega,Oslo,11.0,5.0,1996.0,D,2.290 €,NaN,True,Grup B,Senior analyst,0.0,True,80000.0,0.5,6.5
1005,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Identificación y eliminación de registros completamente vacíos

Se detecta un único registro completamente vacío. Al no contener información en ninguna de sus variables, se considera un registro irrecuperable y se elimina del conjunto de datos, ya que no es posible imputar sus valores de forma fiable.

In [26]:
# Buscar registros completamente vacíos

filas_vacias = df[df.isnull().all(axis=1)]

if filas_vacias.empty:

    print("No se han encontrado registros completamente vacíos.")

else:

    print("Se han encontrado los siguientes registros completamente vacíos:\n")

    display(filas_vacias)

    df = df.drop(index=filas_vacias.index)

    print(
        f"\nSe ha eliminado {len(filas_vacias)} "
        "registro completamente vacío."
    )

print("\nDimensiones actuales del dataset:")
display(df.shape)

Se han encontrado los siguientes registros completamente vacíos:



,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,Càrrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
1005,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN



Se ha eliminado 1 registro completamente vacío.

Dimensiones actuales del dataset:


(986, 19)

**Identificación de campos clave con valores nulos**

Se consideran Nom, Cognoms y DNI como campos clave para identificar a cada empleado. Se revisa la existencia de valores nulos en estas variables para detectar registros potencialmente problemáticos. No obstante, la presencia de un valor nulo en alguno de estos campos no implica necesariamente la eliminación del registro, ya que la decisión depende de si la información restante permite continuar utilizándolo en el análisis.

Comprobación de los campos obligatorios

In [27]:
# Campos considerados obligatorios

campos_obligatorios = [
    "Nom",
    "Cognoms",
    "DNI"
]

# Buscar registros con valores nulos en los campos obligatorios

registros_incompletos = df[
    df[campos_obligatorios].isnull().any(axis=1)
]

if registros_incompletos.empty:

    print(
        "No se han encontrado registros con valores nulos "
        "en los campos obligatorios."
    )

else:

    print(
        "Se han encontrado los siguientes registros con valores "
        "nulos en los campos obligatorios:\n"
    )

    display(registros_incompletos)

No se han encontrado registros con valores nulos en los campos obligatorios.


Antes de realizar las validaciones, se definen las reglas que debe cumplir cada tipo de dato. Estas reglas establecen los rangos permitidos para las variables numéricas, las categorías válidas para las variables de texto y las columnas que deben contener valores booleanos. Posteriormente, estas definiciones se utilizarán para comprobar la calidad de los datos del dataset.

In [28]:
# Reglas de validación

ENTEROS = {
    "Dia de Naixement": (1, 31),
    "Mes de Naixement": (1, 12),
    "Any de Naixement": (1900, 2026),
    "Salari mensual": (0, None),
    "Nombre_fills": (0, None),
    "Km_anuals": (0, None),
}

FLOATS = {
    "Consum_mitja_L_100km": (0, None),
    "Temperatura_mitjana_ciutat": (-273.15, None),
}

STRINGS = {
    "Gènere": {"H", "D", "Home", "Dona"},
    "Grup Professional": {
        "Grup A",
        "Grup B",
        "Grup C",
        "Grup D",
    },
}

BOOLEANOS = [
    "Te_cotxe",
    "Fills",
    "No Fills",
]


Valoracion de tipos de datos y rangos permitidos

En este bloque se valida que los datos almacenados en cada columna cumplan con el formato esperado. Para ello se definen reglas según el tipo de variable:

Variables enteras (ENTEROS): se comprueba que los valores sean numéricos y se encuentren dentro de un rango válido.
Variables decimales (FLOATS): se verifica que sean números decimales y respeten los límites establecidos.
Variables categóricas (STRINGS): se comprueba que el valor pertenezca al conjunto de categorías permitidas.

Cada incidencia detectada se almacena en un diccionario (errores), permitiendo identificar todos los problemas de cada registro antes de decidir qué acciones tomar sobre ellos.

In [29]:
errores = defaultdict(list)

# Validación de variables enteras

for columna, (minimo, maximo) in ENTEROS.items():

    valores = pd.to_numeric(df[columna], errors="coerce")

    # No es un número entero
    mask = valores.isna()

    for indice in df.index[mask]:
        errores[indice].append(
            f"{columna}: '{df.at[indice, columna]}' no es un número entero."
        )

    # Menor que el mínimo
    if minimo is not None:

        mask = valores.notna() & (valores < minimo)

        for indice in df.index[mask]:
            errores[indice].append(
                f"{columna}: '{df.at[indice, columna]}' es menor que {minimo}."
            )

    # Mayor que el máximo
    if maximo is not None:

        mask = valores.notna() & (valores > maximo)

        for indice in df.index[mask]:
            errores[indice].append(
                f"{columna}: '{df.at[indice, columna]}' es mayor que {maximo}."
            )

In [30]:
# Floats.
for columna, (minimo, maximo) in FLOATS.items():

    valores = pd.to_numeric(df[columna], errors="coerce")

    mask = valores.isna()

    for indice in df.index[mask]:
        errores[indice].append(
            f"{columna}: '{df.at[indice, columna]}' no es un número decimal."
        )

    if minimo is not None:

        mask = valores.notna() & (valores < minimo)

        for indice in df.index[mask]:
            errores[indice].append(
                f"{columna}: '{df.at[indice, columna]}' es menor que {minimo}."
            )

    if maximo is not None:

        mask = valores.notna() & (valores > maximo)

        for indice in df.index[mask]:
            errores[indice].append(
                f"{columna}: '{df.at[indice, columna]}' es mayor que {maximo}."
            )

In [31]:
# Strings.
for columna, permitidos in STRINGS.items():

    # No es una cadena
    mask = ~df[columna].map(lambda x: isinstance(x, str))

    for indice in df.index[mask]:
        errores[indice].append(
            f"{columna}: '{df.at[indice, columna]}' no es una cadena de texto."
        )

    # Valor no permitido
    mask = (
        df[columna].map(lambda x: isinstance(x, str))
        & ~df[columna].isin(permitidos)
    )

    for indice in df.index[mask]:
        errores[indice].append(
            f"{columna}: '{df.at[indice, columna]}' "
            f"no pertenece a {sorted(permitidos)}."
        )

In [32]:
#Booleanos.
for columna in BOOLEANOS:
    mask = ~df[columna].map(lambda x: isinstance(x, bool))
    for indice in df.index[mask]:
        errores[indice].append(
            f"{columna}: '{df.at[indice, columna]}' no es un valor booleano."
        )

In [33]:
#DNI.
dni = (
    df["DNI"]
    .astype("string")
    .str.strip()
    .str.upper()
)

mask = ~dni.str.fullmatch(r"\d{8}[A-Z]", na=False)

for indice in df.index[mask]:
    errores[indice].append(
        f"DNI: '{df.at[indice, 'DNI']}' no tiene el formato correcto."
    )

In [34]:
#Fills y no fills.
mask = (
    df["Fills"].notna()
    & df["No Fills"].notna()
    & (df["Fills"] == df["No Fills"])
)

for indice in df.index[mask]:

    errores[indice].append(
        f"Fills={df.at[indice,'Fills']} y "
        f"No Fills={df.at[indice,'No Fills']} "
        "deberían ser valores opuestos."
    )

In [35]:
print(f"Se han detectado errores en {len(errores)} registros.")
for fila, lista in errores.items():

    print(f"\nFila {fila}")

    for error in lista:
        print(f"  • {error}")

Se han detectado errores en 986 registros.

Fila 39
  • Dia de Naixement: '0.0' es menor que 1.
  • Mes de Naixement: '99.0' es mayor que 12.
  • Any de Naixement: '3000.0' es mayor que 2026.
  • Salari mensual: 'dos mil' no es un número entero.
  • Km_anuals: 'nan' no es un número entero.
  • Consum_mitja_L_100km: 'nan' no es un número decimal.
  • Temperatura_mitjana_ciutat: 'nan' no es un número decimal.
  • Te_cotxe: 'nan' no es un valor booleano.
  • Fills: 'nan' no es un valor booleano.

Fila 47
  • Dia de Naixement: '-5.0' es menor que 1.
  • Mes de Naixement: '0.0' es menor que 1.
  • Any de Naixement: '3000.0' es mayor que 2026.
  • Salari mensual: '1939 euros' no es un número entero.
  • Consum_mitja_L_100km: 'nan' no es un número decimal.
  • Temperatura_mitjana_ciutat: 'nan' no es un número decimal.
  • No Fills: 'nan' no es un valor booleano.

Fila 53
  • Dia de Naixement: '0.0' es menor que 1.
  • Mes de Naixement: '0.0' es menor que 1.
  • Salari mensual: '1,866 €' no es

Tratamiento de los registros con errores

## Eliminación de registros con información irrecuperable

Tras aplicar las validaciones sobre los tipos de datos, rangos y valores permitidos, se revisan los registros que presentan incidencias.

Se observa que el registro con índice **1006** acumula múltiples errores que afectan a campos esenciales del dataset, como el DNI, la fecha de nacimiento, variables numéricas y variables categóricas. Además, contiene valores imposibles o claramente inválidos (por ejemplo, fechas fuera de rango, identificadores incorrectos y valores sin significado para el análisis).

Debido a que las inconsistencias afectan a gran parte de la información del registro, no es posible corregirlo ni imputar sus valores de forma fiable sin introducir suposiciones. Por este motivo, el registro se considera **información irrecuperable** y se elimina del conjunto de datos.

In [36]:
# Mostrar el registro antes de eliminarlo

print("Registro con información irrecuperable:")

display(df.loc[[1006]])

# Eliminar el registro

df = df.drop(index=1006)

print("\nSe ha eliminado el registro 1006 por contener información irrecuperable.")

print("\nDimensiones finales del dataset:")

display(df.shape)

Registro con información irrecuperable:


,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,Càrrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
1006,???,###,XXXX,Narnia,Desconeguda,-99.0,99.0,3000.0,??,molts,NaN,NaN,Z,???,NaN,NaN,NaN,NaN,NaN



Se ha eliminado el registro 1006 por contener información irrecuperable.

Dimensiones finales del dataset:


(985, 19)


## Conclusiones

Tras realizar las validaciones básicas del dataset, se han aplicado los siguientes criterios de calidad:

- Se han corregido los problemas de codificación detectados en los nombres de las columnas y en los valores de texto.
- Se han eliminado las filas duplicadas exactas.
- Se ha comprobado la unicidad del DNI, eliminando los registros duplicados que no podían considerarse fiables.
- Se ha identificado y eliminado el registro con índice **1005**, al tratarse de una fila completamente vacía y, por tanto, de información irrecuperable.
- Se han validado los tipos de datos, los rangos permitidos y los valores de las principales variables, detectando tanto valores fuera de los límites establecidos como formatos inconsistentes y distintas representaciones de un mismo valor categórico.
- Finalmente, se ha eliminado el registro con índice **1006**, ya que presentaba múltiples inconsistencias en campos esenciales del dataset que impedían garantizar la fiabilidad de la información.

Aunque se han detectado otras incidencias puntuales, como formatos numéricos inconsistentes o valores categóricos pendientes de normalización, estas no invalidan los registros y podrán resolverse durante la fase de transformación y preparación de los datos del siguiente ejercicio.

Tras estas decisiones, el dataset se considera suficientemente consistente para continuar con las tareas de limpieza, transformación y posterior análisis exploratorio.

El dataset obtenido tras estas validaciones constituye la base sobre la que se realizarán las siguientes fases del análisis exploratorio y la preparación de los datos.

In [37]:
print("Dimensiones finales del dataset:")

display(df.shape)

print("\nPrimeras filas del dataset limpio:")

display(df.head())

Dimensiones finales del dataset:


(985, 19)


Primeras filas del dataset limpio:


,Nom,Cognoms,DNI,País d'origen,Ciutat,Dia de Naixement,Mes de Naixement,Any de Naixement,Gènere,Salari mensual,Fills,No Fills,Grup Professional,Càrrec,Nombre_fills,Te_cotxe,Km_anuals,Consum_mitja_L_100km,Temperatura_mitjana_ciutat
0,Joana,Gil Navarro,66722344X,Espanya,Valladolid,23.0,3.0,1958.0,Dona,"1,469 €",True,NaN,Grup B,Cap de projecte,3.0,True,32108.0,25.0,10.1
1,Marc,Muñoz,48840994W,Espanya,Alacant,8.0,11.0,1960.0,H,"2,718 €",NaN,True,Grup C,Senior analyst,0.0,True,19496.0,10.4,18.7
2,Noa,Serra,14308421X,Espanya,Alacant,27.0,4.0,1961.0,D,1358 euros,True,NaN,Grup A,Tècnic IT,4.0,NaN,NaN,NaN,16.7
3,Pol,Gil,58586340F,Espanya,Sevilla,13.0,10.0,1985.0,H,"1,478 €",True,NaN,Grup B,Data Analyst,2.0,False,NaN,NaN,18.3
4,David,Vila,82070937P,Espanya,Bilbao,29.0,11.0,1965.0,Dona,"1,284 €",NaN,True,Grup B,Administratiu,NaN,True,NaN,11.8,13.1


Dataset final

In [38]:
print(f"El dataset final contiene {df.shape[0]} registros y {df.shape[1]} variables.")

El dataset final contiene 985 registros y 19 variables.


Exercici 3: Transformacions necessàries i preparació del dataset

 Objectiu: preparar el dataset perquè sigui coherent, analitzable i estructuralment correcte abans de realitzar qualsevol anàlisi descriptiva.
En aquest exercici hauràs d'aplicar transformacions bàsiques però essencials, orientades a millorar la qualitat de les dades i facilitar l'anàlisi posterior.

Aspectes a revisar i transformar:

» Els valors categòrics estan normalitzats i són consistents?

» Els valors nuls són coherents amb la informació disponible?

» Hi ha redundància de variables que pugui generar problemes analítics?

» Les variables contextuals (com temperatura, salari) estan completes i preparades per a l'anàlisi?

Concretament, hauràs de:

» Normalitzar els valors categòrics per evitar ambigüitats (per exemple, valors que representen el mateix concepte però estan escrits de manera diferent).
» Revisar la coherència dels valors nuls, assegurant que:
Les persones que tenen quilometratge anual i consum mitjà associats disposen de cotxe. En els casos en què consti consum mitjà però no quilometratge anual, caldrà decidir com resoldre aquesta incoherència de manera justificada.
Les persones que tenen un valor en el nombre de fills efectivament tenen fills.
» Unificar la informació sobre fills en una única columna, ja que mantenir diverses columnes amb informació complementària no és òptim i pot generar problemes analítics (redundància o multicolinealitat).
» Validar la temperatura mitjana. En cas que aquesta informació falti, hauràs d'imputar-la utilitzant un criteri raonable (per ciutat o per un nivell superior justificat).
» Revisar i imputar els salaris faltants, aplicant criteris analítics coherents (com ara càrrec, grup professional, ciutat, gènere o franja d'edat).

Bloque 1
Normalizar los valores categoricos para evitar embiguedades y garantizar consistencia en las variables

Antes de normalizar los valores categóricos, es necesario conocer qué valores distintos existen realmente en cada variable para identificar posibles inconsistencias o diferentes representaciones de un mismo concepto.

**Revisar las categorías de las variables categóricas**

In [39]:
for col in df.select_dtypes(include="object").columns:
    print(f"\n--- {col} ---")
    display(
        df[col]
        .value_counts(dropna=False)
        .rename_axis("Categoria")
        .reset_index(name="Frecuencia")
    )


--- Nom ---


,Categoria,Frecuencia
0,Adrià,34
1,Aina,29
2,David,28
3,Sergio,28
4,Alba,27
...,...,...
97,Kristian,1
98,Finn,1
99,Emma,1
100,Magnus,1



--- Cognoms ---


,Categoria,Frecuencia
0,Ferrer,21
1,García,19
2,Díaz,18
3,Moreno,18
4,Costa,17
...,...,...
461,Fernández Moreno,1
462,Ribera Costa,1
463,Vidal Casals,1
464,Lefebvre Moreau,1



--- DNI ---


,Categoria,Frecuencia
0,62042563B,1
1,66722344X,1
2,48840994W,1
3,14308421X,1
4,58586340F,1
...,...,...
980,66851377P,1
981,31172421Z,1
982,52169044F,1
983,82772208R,1



--- País d'origen ---


,Categoria,Frecuencia
0,Espanya,700
1,Itàlia,108
2,França,86
3,Alemanya,48
4,Noruega,43



--- Ciutat ---


,Categoria,Frecuencia
0,Palma,57
1,Barcelona,54
2,Màlaga,53
3,Murcia,52
4,Sevilla,48
5,Zaragoza,47
6,Bilbao,46
7,Las Palmas de Gran Canaria,45
8,A Coruña,45
9,Tarragona,44



--- Gènere ---


,Categoria,Frecuencia
0,D,317
1,H,274
2,Dona,117
3,F,107
4,Home,50
5,NaN,39
6,M,34
7,A,24
8,NC,23



--- Salari mensual ---


,Categoria,Frecuencia
0,n/d,14
1,mil vuit-cents,9
2,dos mil,7
3,tres mil,6
4,"1,644 €",4
...,...,...
836,1479€,1
837,"1,840 €",1
838,"1,338 €",1
839,"2,855 €",1



--- Fills ---


,Categoria,Frecuencia
0,NaN,574
1,True,411



--- No Fills ---


,Categoria,Frecuencia
0,True,572
1,NaN,413



--- Grup Professional ---


,Categoria,Frecuencia
0,Grup A,456
1,Grup B,350
2,Grup C,139
3,Grup D,40



--- Càrrec ---


,Categoria,Frecuencia
0,Data Analyst,139
1,Tècnic IT,128
2,Senior analyst,122
3,Responsable de vendes,121
4,Analista,112
5,Analista junior,106
6,Administratiu,102
7,Cap de projecte,93
8,Project lead,17
9,Data analyst,17



--- Te_cotxe ---


,Categoria,Frecuencia
0,NaN,342
1,True,323
2,False,320


Comprobar los tipos de datos dentro de las variables categóricas

Aunque una columna tenga tipo object, esto no garantiza que todos sus valores sean cadenas de texto. Se comprueba el tipo de los datos almacenados en cada variable para identificar posibles mezclas de tipos (por ejemplo, cadenas, booleanos o valores nulos) y aplicar posteriormente las transformaciones únicamente cuando corresponda.

In [47]:
columnas_categoricas = df.select_dtypes(include="object").columns

for col in columnas_categoricas:
    print(f"\n--- {col} ---")
    print(df[col].apply(type).value_counts())


--- Nom ---
Nom
<class 'str'>    985
Name: count, dtype: int64

--- Cognoms ---
Cognoms
<class 'str'>    985
Name: count, dtype: int64

--- DNI ---
DNI
<class 'str'>    985
Name: count, dtype: int64

--- País d'origen ---
País d'origen
<class 'str'>    985
Name: count, dtype: int64

--- Ciutat ---
Ciutat
<class 'str'>    985
Name: count, dtype: int64

--- Gènere ---
Gènere
<class 'str'>      946
<class 'float'>     39
Name: count, dtype: int64

--- Salari mensual ---
Salari mensual
<class 'str'>    985
Name: count, dtype: int64

--- Fills ---
Fills
<class 'float'>    574
<class 'bool'>     411
Name: count, dtype: int64

--- No Fills ---
No Fills
<class 'bool'>     572
<class 'float'>    413
Name: count, dtype: int64

--- Grup Professional ---
Grup Professional
<class 'str'>    985
Name: count, dtype: int64

--- Càrrec ---
Càrrec
<class 'str'>    985
Name: count, dtype: int64

--- Te_cotxe ---
Te_cotxe
<class 'bool'>     643
<class 'float'>    342
Name: count, dtype: int64


Se eliminan los espacios al inicio y al final de las cadenas de texto y se convierten todos los valores a minúsculas. Esta transformación evita que diferencias de formato generen categorías distintas que representen el mismo concepto. La comprobación previa de los tipos de datos permite aplicar la transformación únicamente a los valores de tipo str, preservando el resto de tipos de datos.

In [51]:
df[columnas_categoricas] = df[columnas_categoricas].apply(
    lambda col: col.map(
        lambda x: x.strip().lower() if isinstance(x, str) else x
    )
)

Revisar nuevamente las categorias

In [52]:
for col in df.select_dtypes(include="object").columns:
    print(f"\n--- {col} ---")
    display(
        df[col]
        .value_counts(dropna=False)
        .rename_axis("Categoria")
        .reset_index(name="Frecuencia")
    )


--- Nom ---


,Categoria,Frecuencia
0,adrià,34
1,aina,29
2,david,28
3,sergio,28
4,alba,27
...,...,...
97,kristian,1
98,finn,1
99,emma,1
100,magnus,1



--- Cognoms ---


,Categoria,Frecuencia
0,ferrer,21
1,garcía,19
2,díaz,18
3,moreno,18
4,costa,17
...,...,...
461,fernández moreno,1
462,ribera costa,1
463,vidal casals,1
464,lefebvre moreau,1



--- DNI ---


,Categoria,Frecuencia
0,62042563b,1
1,66722344x,1
2,48840994w,1
3,14308421x,1
4,58586340f,1
...,...,...
980,66851377p,1
981,31172421z,1
982,52169044f,1
983,82772208r,1



--- País d'origen ---


,Categoria,Frecuencia
0,espanya,700
1,itàlia,108
2,frança,86
3,alemanya,48
4,noruega,43



--- Ciutat ---


,Categoria,Frecuencia
0,palma,57
1,barcelona,54
2,màlaga,53
3,murcia,52
4,sevilla,48
5,zaragoza,47
6,bilbao,46
7,las palmas de gran canaria,45
8,a coruña,45
9,tarragona,44



--- Gènere ---


,Categoria,Frecuencia
0,d,317
1,h,274
2,dona,117
3,f,107
4,home,50
5,NaN,39
6,m,34
7,a,24
8,nc,23



--- Salari mensual ---


,Categoria,Frecuencia
0,n/d,14
1,mil vuit-cents,9
2,dos mil,7
3,tres mil,6
4,"1,644 €",4
...,...,...
836,1479€,1
837,"1,840 €",1
838,"1,338 €",1
839,"2,855 €",1



--- Fills ---


,Categoria,Frecuencia
0,NaN,574
1,True,411



--- No Fills ---


,Categoria,Frecuencia
0,True,572
1,NaN,413



--- Grup Professional ---


,Categoria,Frecuencia
0,grup a,456
1,grup b,350
2,grup c,139
3,grup d,40



--- Càrrec ---


,Categoria,Frecuencia
0,data analyst,156
1,tècnic it,128
2,senior analyst,122
3,responsable de vendes,121
4,analista,112
5,cap de projecte,106
6,analista junior,106
7,administratiu,102
8,project lead,17
9,analyst,9



--- Te_cotxe ---


,Categoria,Frecuencia
0,NaN,342
1,True,323
2,False,320


Antes de unificar categorias equivalentes, como en el caso de la columna 'genere', procedo a verificar si hay algun patron que demuestre la equivalencia con los valores asignados.

Al agrupar por nombre y edad puedes constatar que no corresponden todos los nombres con el genero correspondiente. Por lo que hay un problema de consistencia.

In [57]:
df.groupby("Gènere")[["Nom", "Gènere"]].head()

,Nom,Gènere
0,joana,dona
1,marc,h
2,noa,d
3,pol,h
4,david,dona
5,alexia,d
6,david,d
7,carmen,h
540,nora,d
9,alba,h


In [58]:
for valor in df["Gènere"].dropna().unique():
    print(f"\n=== {valor} ===")
    print(df.loc[df["Gènere"] == valor, "Nom"].head(10).tolist())


=== dona ===
['joana', 'david', 'sergio', 'paul', 'pau', 'marina', 'giorgio', 'noa', 'xavier', 'pau']

=== h ===
['marc', 'pol', 'carmen', 'alba', 'guillem', 'aina', 'xavier', 'carmen', 'chloé', 'adrià']

=== d ===
['noa', 'alexia', 'david', 'nora', 'emilia', 'sigrid', 'adrià', 'riccardo', 'valentina', 'camille']

=== f ===
['joana', 'laia', 'aina', 'sergio', 'marta', 'clara', 'aina', 'alba', 'pol', 'claudia']

=== home ===
['sara', 'nil', 'aina', 'víctor', 'irene', 'pol', 'anna', 'alessandro', 'irene', 'marina']

=== nc ===
['antoine', 'carlos', 'elena', 'thomas', 'simone', 'noa', 'núria', 'anna', 'gabriele', 'anna']

=== m ===
['adrià', 'francesca', 'carlos', 'jordi', 'alexia', 'noa', 'adrià', 'chiara', 'marta', 'oriol']

=== a ===
['sara', 'víctor', 'camille', 'marina', 'david', 'laia', 'pol', 'marta', 'alexia', 'alexia']


Tras revisar la variable Gènere, se observa que una misma categoría aparece representada mediante diferentes codificaciones (home, h y m; dona, d y f). Para homogeneizar la variable, se unifican estas abreviaturas utilizando las categorías completas home y dona.

No obstante, se detectan registros aparentemente inconsistentes (por ejemplo, nombres tradicionalmente femeninos asociados a h o masculinos asociados a dona). Dado que el dataset no proporciona información suficiente para determinar el valor correcto, no se corrigen estos casos, evitando introducir modificaciones no justificadas. Los valores a y nc se mantienen sin cambios por el mismo motivo.

In [ ]:
map_genero = {
    "h": "home",
    "m": "home",
    "d": "dona",
    "f": "dona"
}

df["Gènere"] = df["Gènere"].replace(map_genero)

**Bloque 2. revisar la coherencia de los valores nulos. Coherencia de la información sobre el coche**

Explorar las variables relacionadas con el coche

In [66]:
print(df.columns.tolist())

['Nom', 'Cognoms', 'DNI', "País d'origen", 'Ciutat', 'Dia de Naixement', 'Mes de Naixement', 'Any de Naixement', 'Gènere', 'Salari mensual', 'Fills', 'No Fills', 'Grup Professional', 'Càrrec', 'Nombre_fills', 'Te_cotxe', 'Km_anuals', 'Consum_mitja_L_100km', 'Temperatura_mitjana_ciutat']


Bloque 3. Coherencia de los datos sobre hijos -- Hi ha redundància de variables que pugui generar problemes analítics?   ej: fills/no fills

Bloque 4. Unificación de la información sobre hijos - modificacion de variables desaparece fills/no fills y se queda fills

Bloque 5. Imputación de la temperatura
Validar la temperatura mitjana.

Aquí analizaremos:

ciudades repetidas
temperaturas por ciudad
imputar por ciudad
si una ciudad no tiene temperatura, usar provincia o media global (según lo que haya en el dataset).

Bloque 6. Imputación del salario

Responde a:

Revisar i imputar els salaris faltants.

Aquí habrá que decidir un criterio.

Probablemente:

Grupo profesional.
Ciudad.
Edad.
Género (solo si realmente aporta información).

Bloque 7. Comprobación final